Use right calling convention for JNI calls

radekdoulik opened the pull request:
Fixes (partially?) https://github.com/xamarin/java.interop/issues/403

Make sure we use the right calling convention by using JNICALL.

Do not force `CallingConvention.Cdecl` in pinvokes, as these are
different on Windows. Rather stay with the platform defaults.

jonpryor commented on src/java-interop/java-interop-jvm.c:
```diff
@@ -6,8 +6,8 @@
 
-typedef int (*java_interop_JNI_CreateJavaVM_fptr) (JavaVM **p_vm, void **p_env, void *vm_args);
-typedef int (*java_interop_JNI_GetCreatedJavaVMs_fptr) (JavaVM **vmBuf, int bufLen, int *nVMs);
+typedef int (JNICALL *java_interop_JNI_CreateJavaVM_fptr) (JavaVM **p_vm, void **p_env, void *vm_args);
```
This looks important for part of https://github.com/xamarin/xamarin-android/pull/2525 ...

jonpryor replied:
The changes to this file are the *only* changes that appear to be required.

jonpryor commented:
> Do not force CallingConvention.Cdecl in pinvokes, as these are different on Windows. Rather stay with the platform defaults.

This (unfortunately) doesn't mean what you think (hope) it means.

*By default*, MSVC on Windows generates cdecl calls.  Meaning this:

```
int
java_interop_gc_bridge_get_gref_count (JavaInteropGCBridge *bridge)
{
    ...
```

on Windows will generate `java_interop_gc_bridge_get_gref_count()` with a *`cdecl`* calling convention, *not* `__stdcall`.

Meaning *as-is*, this PR will *further* break things on Windows, because the P/Invoke declaration for `java_interop_gc_bridge_get_gref_count()` now implicitly becomes stdcall on Windows, while the C library will *remain* cdecl, which will screw up the stack.

Ditto the generated `jni.c`: we currently generate -- and this PR doesn't alter -- the following:

```c
JI_API jint
java_interop_jnienv_get_version (JNIEnv *env)
{
	jint _r_ = (*env)->GetVersion (env);
	return _r_;
}
```

`java_interop_jnienv_get_version()` will have a cdecl calling convention on Windows, while the change to `build-tools/jnienv-gen/Generator.cs` would alter the C# side to use the platform default -- stdcall on Windows -- which, again, will screw up the stack.

In short, this PR is *not* what we want, not as is anyway.

radekdoulik commented:
That's not what I saw on Windows. The PInvoke defaults work OK without explicitly specifying Cdecl.

Context: https://github.com/xamarin/java.interop/issues/403#issuecomment-447023389

radekdoulik commented:
I have tested that the following `DllImport`s work with and without explicitly adding `Cdecl`.

                [DllImport (JavaInteropLib, CharSet=CharSet.Ansi)]
                internal static extern int java_interop_jvm_load (string path);

                [DllImport (JavaInteropLib, CharSet=CharSet.Ansi)]
                internal static extern int java_interop_jvm_create (out IntPtr javavm, out IntPtr jnienv, ref JavaVMInitArgs args);

After the discussion we decided to keep the Cdecl's around, ideally we should specify the calling conventions in the native code as well. We can do that later if we run into issues, as Jonathan wants to avoid bigger changes.

So I have updated the patch to use Cdecl on managed side and added it to few places for consistency.
